### PR TITLE
Jsonnet config

### PR DESCRIPTION
**Pull request by mmsqe**

### CODE REVIEW

File: pystarport/cluster.py [hunk 1]
```diff
@@ -924,7 +924,11 @@ def init_cluster(
     cmd=None,
     gen_compose_file=False,
 ):
-    config = expand_yaml(config_path, dotenv)
+    extension = config_path.rsplit('.', 1)[1]
```
Comment (yihuang) [inline pystarport/cluster.py:927]: Use [.suffix](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.suffix) of the `Path` object.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -22,6 +22,7 @@ jobs:
   tests:
     runs-on: ubuntu-latest
     steps:
+      - uses: zendesk/setup-jsonnet@v10
```
Comment (yihuang) [inline .github/workflows/test.yml:25]: is it necessary now, since we have included in the `pyproject.yaml`?

File: .gitignore [hunk 3]
```diff
@@ -138,4 +138,7 @@ dmypy.json
 cython_debug/
 
 # VS Code configuration
-.vscode
\ No newline at end of file
+.vscode
+
+# integration yamls
+pystarport/tests/test_expansion/*.yaml
```
Comment (yihuang) [inline .gitignore:144]: I think no need to remove the yaml test cases? since we still support them.

File: Makefile [hunk 4]
```diff
@@ -1,5 +1,5 @@
 test:
-	@(cd pystarport/tests && poetry run pytest)
+	@(cd pystarport/tests && ./test_expansion/generate-test-yamls && poetry run pytest)
```
Comment (yihuang) [inline Makefile:2]: I see, you converted the yaml to jsonnet and generate it on the fly, what's the reason of doing this?

File: pystarport/expansion.py [hunk 5]
```diff
@@ -44,12 +46,34 @@ def _expand(value, variables):
...
+    if dotenv is not None:
+        if "dotenv" in config:
+            _ = config.pop("dotenv", {})  # remove dotenv field if exists
+    elif "dotenv" in config:
+        dotenv = config.pop("dotenv", {})  # pop dotenv field if exists
+
+    if dotenv:
+        if not isinstance(dotenv, str):
+            raise ValueError(f"Invalid value passed to dotenv: {dotenv}")
+        env_path = path.parent.joinpath(dotenv)
```
Comment (yihuang) [inline pystarport/expansion.py:61]:
```suggestion
        env_path = path.parent / dotenv
```

File: .gitignore (hunk 3 above)
Comment (mmsqe) [reply .gitignore]: just to keep [yaml](https://github.com/crypto-com/pystarport/issues/50) as default

File: Makefile (hunk 4 above)
Comment (yihuang) [reply Makefile:2]: The #50 is more or less fixed by yaml include support I think, this PR is mainly to add more expressiveness by support jsonnet.

**Review by yihuang — CHANGES_REQUESTED** [review]
Looks good overall, just think we don't need change the existing yaml test cases, just add new ones.

File: pystarport/tests/test_expansion/cronos_has_dotenv.yaml [hunk 6]
```diff
@@ -1,27 +1,52 @@
-include: !include base.yaml
```
[the remark this answers is not in the document]
Comment (mmsqe) [reply pystarport/tests/test_expansion/cronos_has_dotenv.yaml]: revert the yaml diff generated by jsonnet, only sort change ensure by cross check test

File: pystarport/cluster.py [hunk 7]
```diff
@@ -924,7 +924,11 @@ def init_cluster(
     cmd=None,
     gen_compose_file=False,
 ):
-    config = expand_yaml(config_path, dotenv)
+    extension = PurePosixPath(config_path).suffix
```
Comment (yihuang) [inline pystarport/cluster.py:927]:
```suggestion
    extension = Path(config_path).suffix
```

File: pystarport/cluster.py [hunk 8]
```diff
@@ -9,7 +9,7 @@
 import sys
 import threading
 import time
-from pathlib import Path
+from pathlib import Path, PurePosixPath
```
Comment (yihuang) [inline pystarport/cluster.py:12]:
```suggestion
from pathlib import Path
```

**Review by yihuang — APPROVED** [review]
LGTM.
Only one minor suggestion is use [parametrized](https://docs.pytest.org/en/6.2.x/parametrize.html) feature of pytest to structure the yaml/jsonnet test case.

File: pystarport/tests/test_expansion/test_expansion.py [hunk 9]
```diff
@@ -4,31 +4,55 @@
...
+    bases = []
+    cronos_has_dotenvs = []
+    cronos_no_dotenvs = []
+    cronos_has_posix_no_dotenvs = []
     parent = Path(__file__).parent
-    base = parent / "base.yaml"
-    cronos_has_dotenv = parent / "cronos_has_dotenv.yaml"
-    cronos_no_dotenv = parent / "cronos_no_dotenv.yaml"
-    cronos_has_posix_no_dotenv = parent / "cronos_has_posix_no_dotenv.yaml"
+    for type in [".yaml", ".jsonnet"]:
```
[the remark this answers is not in the document]
Comment (mmsqe) [reply pystarport/tests/test_expansion/test_expansion.py:24]: need call function, cant use parameter ar

[the remark this answers is not in the document]
Comment (yihuang) [reply pystarport/tests/test_expansion/test_expansion.py:24]: it'll run your test case once for each parameter automatically.

[the remark this answers is not in the document]
Comment (mmsqe) [reply pystarport/tests/test_expansion/test_expansion.py]: need [check](https://github.com/crypto-com/pystarport/pull/52/commits/8ca031f03e7003271e927df629a3978bf744e421) to know which function to call

**Pull request merged**